src/bin/nydus-image/deduplicate.rs

```diff
@@ -0,0 +1,496 @@
+use std::sync::{Arc, Mutex};
+
+pub trait Database {
+    /// Creates a new chunk in the database.
+    fn create_chunk(&self) -> Result<(), ()>;
+
+    /// Creates a new blob in the database.
+    fn create_blob(&self) -> Result<(), ()>;
+
+    /// Inserts chunk information into the database.
+    fn insert_chunk(&self, chunk_info: &ChunkTable) -> Result<(), ()>;
+
+    /// Inserts blob information into the database.
+    fn insert_blob(&self, blob_info: &BlobTable) -> Result<(), ()>;
+
+    /// Retrieves all chunk information from the database.
+    fn get_chunks(&self) -> Result<Vec<ChunkTable>, ()>;
+
+    /// Retrieves all blob information from the database.
+    fn get_blobs(&self) -> Result<Vec<BlobTable>, ()>;
+}
+
+pub struct SqliteDatabase {
+    conn: Arc<Mutex<Connection>>,
+}
+
+impl SqliteDatabase {
+    pub fn new(database_path: &str) -> Result<Self, rusqlite::Error> {
+        // Delete the database file if it exists.
+        if let Ok(metadata) = fs::metadata(database_path) {
+            if metadata.is_file() {
+                if let Err(err) = fs::remove_file(database_path) {
+                    warn!(
+                        "Warning: Unable to delete existing database file: {:?}.",
+                        err
+                    );
+                }
+            }
+        }
+
+        // Attempt to open a new SQLite connection.
+        let conn = Connection::open(database_path)?;
+
+        Ok(Self {
+            conn: Arc::new(Mutex::new(conn)),
+        })
+    }
+}
+
+impl Database for SqliteDatabase {
+    fn create_chunk(&self) -> Result<(), ()> {
+        let conn = self.conn.lock().unwrap();
+        ChunkTable::create(&conn).map_err(|_| ())
+    }
+
+    fn create_blob(&self) -> Result<(), ()> {
+        let conn = self.conn.lock().unwrap();
+        BlobTable::create(&conn).map_err(|_| ())
+    }
+
+    fn insert_chunk(&self, chunk: &ChunkTable) -> Result<(), ()> {
+        let conn = self.conn.lock().unwrap();
+        ChunkTable::insert(&conn, chunk).map_err(|_| ())
+    }
+
+    fn insert_blob(&self, blob: &BlobTable) -> Result<(), ()> {
+        let conn = self.conn.lock().unwrap();
+        BlobTable::insert(&conn, blob).map_err(|_| ())
+    }
+
+    fn get_chunks(&self) -> Result<Vec<ChunkTable>, ()> {
+        let conn = self.conn.lock().unwrap();
+        ChunkTable::list_all(&conn).map_err(|_| ())
+    }
+
+    fn get_blobs(&self) -> Result<Vec<BlobTable>, ()> {
+        let conn = self.conn.lock().unwrap();
+        BlobTable::list_all(&conn).map_err(|_| ())
+    }
+}
+
+pub struct Deduplicate<D: Database + Send + Sync> {
+    sb: RafsSuper,
+    db: D,
+}
+
+impl Deduplicate<SqliteDatabase> {
+    pub fn new(
+        bootstrap_path: &Path,
+        config: Arc<ConfigV2>,
+        db_path: &str,
+    ) -> anyhow::Result<Self> {
+        let (sb, _) = RafsSuper::load_from_file(bootstrap_path, config, false)?;
+        let db = SqliteDatabase::new(db_path.strip_prefix('/').unwrap_or(db_path))?;
```

Comment:
Why need `strip_prefix` here? We'd better limit the database path to an absolute path.